cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -1565,6 +1564,103 @@ data MultiSigScript = RequireSignature (Hash PaymentKey)
                     | RequireMOf Int [MultiSigScript]
   deriving (Eq, Show)
 
+instance ToJSON MultiSigScript where
+  toJSON (RequireSignature payKeyHash) = String . Text.decodeUtf8 $ serialiseToRawBytesHex $ payKeyHash
+  toJSON (RequireAnyOf reqSigs) = object [ "any" .= map toJSON reqSigs ]
+  toJSON (RequireAllOf reqSigs) = object [ "all" .= map toJSON reqSigs ]
+  toJSON (RequireMOf reqNum reqSigs) =
+    case reqNum > length reqSigs of
+      False -> object [convertMofN .= map toJSON reqSigs]
+      True -> error $ "mofn: Required number of signatures \
+                      \exceed number of payment key hashes provided. m: "
+                      ++ show reqNum ++ " n: " ++ show (length reqSigs)
+   where
+     convertMofN = Text.pack $ (show reqNum ++ "of" ++ show (length reqSigs))
+
+instance FromJSON MultiSigScript where
+  parseJSON = Aeson.withObject "MultiSigScript" $ \obj -> do
+    -- Parse "all" multisig objects
+    allReqHashes <- case HMS.lookup "all" obj of
+                       Just (Array vecAllReqHashes) ->
+                         return . Just . RequireAllOf $ gatherSignatures vecAllReqHashes
+                       _ -> return Nothing
+
+    -- Parse "any" multisig objects
+    anyReqHashes <- case HMS.lookup "any" obj of
+                       Just (Array vecAnyReqHashes) ->
+                         return . Just . RequireAnyOf $ gatherSignatures vecAnyReqHashes
+                       _ -> return Nothing
+
+    -- Parse "mofn" multisig objects
+    mofN <- return $ filterMofN obj
+
+    case (allReqHashes, anyReqHashes, mofN) of
+      (Nothing, Nothing, Nothing) -> fail "Script does not contain \"all\", \"any\" \
+                                          \or \"mofn\" JSON objects."
+      (Just allMss, Nothing, Nothing) -> return allMss
+      (Nothing, Just anyMss, Nothing) -> return anyMss
+      (Nothing, Nothing, Just mOfnMss) -> return mOfnMss
+      (_, _, _) -> fail "Please specify only one type of JSON multisig object."
+
+gatherSignatures :: Vector Value -> [MultiSigScript]
+gatherSignatures anyList =
+  let mPaymentKeyHashes = Vector.map (\val -> RequireSignature <$> filterValue val) anyList
+  in catMaybes $ Vector.toList mPaymentKeyHashes
+ where
+  filterValue :: Value -> Maybe (Hash PaymentKey)
+  filterValue (String hpk) = Just $ convertToHash hpk
+  filterValue _ = Nothing
```

Comment:
Aren't scripts supposed to be able to contain other nested scripts and not just `RequireSignature`s?

Reply:
Discussed on a video call. The consensus is to leave the scripts as simple as possible and adjust if user requirements change.